bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -2439,17 +2433,30 @@ void ReplicaImp::onMessage<AskForCheckpointMsg>(AskForCheckpointMsg *msg) {
   }
 }
 
-void ReplicaImp::startExecution(SeqNum seqNumber,
-                                concordUtils::SpanWrapper &span,
-                                bool askForMissingInfoAboutCommittedItems) {
-  consensus_times_.end(seqNumber);
-  tryToRemovePendingRequestsForSeqNum(seqNumber);  // TODO(LG) Should verify if needed
-  LOG_INFO(CNSUS, "Starting execution of seqNumber:" << seqNumber);
-  if (config_.enablePostExecutionSeparation) {
-    tryToStartOrFinishExecution(askForMissingInfoAboutCommittedItems);
-  } else {
-    executeNextCommittedRequests(span, askForMissingInfoAboutCommittedItems);
+void ReplicaImp::startExecution() {
+  if (isCollectingState() || bftEngine::ControlStateManager::instance().isWedged() ||
+      bftEngine::ControlStateManager::instance().getPruningProcessStatus()) {
+    return;
+  }
+  if (lastCollectedSn % checkpointWindowSize == 0 && main_execution_engine_->isExecuting()) {
+    LOG_INFO(GL, "we must wait for the latest checkpoint creation to finish before continue executing");
+    return;
+  }
+
+  if (!deferredRORequests_.empty()) {
+    LOG_INFO(GL, "prioritize read only reqeusts on top of executions");
+    return;
   }
+  concordUtils::SpanWrapper span;
+  auto pps = collectPrePrepares(span, true);
+  for (auto pp : pps) tryToRemovePendingRequestsForSeqNum(pp->seqNumber());
+  if (pps.empty()) return;
+  LOG_INFO(CNSUS, "Start execution");
```

Comment:
I think it should contain the sequence number of the PP, for debugging purposes.